lms/services/h_api.py

```diff
@@ -0,0 +1,177 @@
+            display_name=user_info["display_name"],
+        )
+
+    def create_user(self, h_user, provider, provider_unique_id):
+        """
+        Create a user in H.
+
+        :param h_user: An `HUser` object to send to H
+        :param provider: The provider (hosting LMS)
+        :param provider_unique_id: The id of the provider
+        """
+        user_data = {
+            "username": h_user.username,
+            "display_name": h_user.display_name,
+            "authority": self._authority,
+            "identities": [
+                {"provider": provider, "provider_unique_id": provider_unique_id,}
+            ],
+        }
+
+        self._api_request("POST", "users", data=user_data)
+
+    def update_user(self, h_user):
+        """
+        Update details for a user in H.
+
+        Currently this only updates the display name.
+
+        :param h_user: An `HUser` object to update
+        """
+        self._api_request(
+            "PATCH",
+            f"users/{h_user.username}",
+            data={"display_name": h_user.display_name},
+        )
+
+    def upsert_user(self, h_user, provider, provider_unique_id):
+        """
+        Create or update a user in H as appropriate.
+
+        This is equivalent to calling `update_user()` then `create_user()` if
+        that fails.
+
+        :param h_user: An `HUser` object to send to H
+        :param provider: The provider (hosting LMS)
+        :param provider_unique_id: The id of the provider
+        """
+        try:
+            self.update_user(h_user)
+        except HAPINotFoundError:
+            self.create_user(h_user, provider, provider_unique_id)
+
+    def create_group(self, group_id, group_name, h_user):
+        """
+        Create a group in H.
+
+        :param group_id: The id of the group
+        :param group_name: The display name for the group
+        :param h_user: An `HUser` to associate with the group
+        """
+        self._api_request(
+            "PUT",
+            f"groups/{group_id}",
+            data={"groupid": group_id, "name": group_name},
+            headers={"X-Forwarded-User": h_user.userid},
+        )
+
+    def update_group(self, group_id, group_name):
+        """
+        Update a group in H.
+
+        Currently this only updates the name of the group.
+
+        :param group_id: The id of the group
+        :param group_name: The display name for the group
+        """
+        self._api_request(
+            "PATCH", f"groups/{group_id}", data={"name": group_name},
+        )
+
+    def add_user_to_group(self, h_user, group_id):
+        """
+        Add the user as a member of the group.
+
+        :param h_user: An `HUser` to associate with the group
+        :param group_id: The id of the group
+        """
+        self._api_request("POST", f"groups/{group_id}/members/{h_user.userid}")
+
+    def _api_request(
+        self, method: str, path: str, data: dict = None, headers: dict = None
+    ):
```

Comment:
All options that were not in use in this class have been removed, simplifying the function. Type declarations have moved to the signature to clean up the docs. I also removed some redundancy in the description describing the `:raise:` portion again.